test/e2e_node/topology_manager_test.go

```diff
@@ -350,27 +501,252 @@ func runTopologyManagerSuiteTests(f *framework.Framework) {
 	waitForContainerRemoval(pod2.Spec.Containers[0].Name, pod2.Name, pod2.Namespace)
 }
 
+func runTopologyManagerPositiveTest(f *framework.Framework, numaNodes, numPods int, cpuAmount, sriovResourceName, deviceAmount string) {
+	var pods []*v1.Pod
+
+	for podID := 0; podID < numPods; podID++ {
+		ctnAttrs := []tmCtnAttribute{
+			{
+				ctnName:       "gu-container",
+				cpuRequest:    cpuAmount,
+				cpuLimit:      cpuAmount,
+				deviceName:    sriovResourceName,
+				deviceRequest: deviceAmount,
+				deviceLimit:   deviceAmount,
+			},
+		}
+
+		podName := fmt.Sprintf("gu-pod-%d", podID)
+		framework.Logf("creating pod %s attrs %v", podName, ctnAttrs)
+		pod := makeTopologyManagerTestPod(podName, numalignCmd, ctnAttrs)
+		pod = f.PodClient().CreateSync(pod)
+		framework.Logf("created pod %s", podName)
+		pods = append(pods, pod)
+	}
+
+	for podID := 0; podID < numPods; podID++ {
+		validatePodAlignment(f, pods[podID], numaNodes)
+	}
+
+	for podID := 0; podID < numPods; podID++ {
+		pod := pods[podID]
+		framework.Logf("deleting the pod %s/%s and waiting for container %s removal",
+			pod.Namespace, pod.Name, pod.Spec.Containers[0].Name)
+		deletePods(f, []string{pod.Name})
+		waitForContainerRemoval(pod.Spec.Containers[0].Name, pod.Name, pod.Namespace)
+	}
+}
+
+func runTopologyManagerNegativeTest(f *framework.Framework, numaNodes, numPods int, cpuAmount, sriovResourceName, deviceAmount string) {
+	ctnAttrs := []tmCtnAttribute{
+		{
+			ctnName:       "gu-container",
+			cpuRequest:    cpuAmount,
+			cpuLimit:      cpuAmount,
+			deviceName:    sriovResourceName,
+			deviceRequest: deviceAmount,
+			deviceLimit:   deviceAmount,
+		},
+	}
+
+	podName := "gu-pod"
+	framework.Logf("creating pod %s attrs %v", podName, ctnAttrs)
+	pod := makeTopologyManagerTestPod(podName, numalignCmd, ctnAttrs)
+
+	pod = f.PodClient().Create(pod)
+	err := e2epod.WaitForPodCondition(f.ClientSet, f.Namespace.Name, pod.Name, "Failed", 30*time.Second, func(pod *v1.Pod) (bool, error) {
+		if pod.Status.Phase != v1.PodPending {
+			return true, nil
+		}
+		return false, nil
+	})
+	framework.ExpectNoError(err)
+	pod, err = f.PodClient().Get(pod.Name, metav1.GetOptions{})
+	framework.ExpectNoError(err)
+
+	if pod.Status.Phase != v1.PodFailed {
+		framework.Failf("pod %s not failed: %v", pod.Name, pod.Status)
+	}
+	if !isTopologyAffinityError(pod) {
+		framework.Failf("pod %s failed for wrong reason: %q", pod.Name, pod.Status.Reason)
+	}
+
+	deletePods(f, []string{pod.Name})
+}
+
+func isTopologyAffinityError(pod *v1.Pod) bool {
+	re := regexp.MustCompile(`Topology.*Affinity.*Error`)
```

Comment:
The negative test was failing in my setup. Debugged it, and could see from older version where it was failing. Code was checking for Topology Affinity Error, from below const definition. I see you have fixed this in latest update. Nice!

topologyError = "Topology Affinity Error" // XXX do we have a proper constant?

Reply:
Sorry for causing you troubles. I'm not very happy with the current test but it's the best I could think of, lacking constants or symbols I could depend on :\

Reply:
Not sure if this is relevant, but just FYI: Changes introduced in https://github.com/kubernetes/kubernetes/pull/87758 mean the error is now "TopologyAffinityError" as opposed to previous "Topology Affinity Error".

Reply:
Ack, I missed this, thanks for pointing it out. This is the reason why it worked before and then failed. Anyway I'd keep the regexp match - lacking better alternatives - because its' a little less fragile than the string matching I had before.